File: sdk/ios/Sources/Causality/Internal/Bridge.swift
import Foundation
import CausalityCore

/// Internal bridge to Go core via JSON
enum Bridge {
    static func initSDK(config: Config) throws {
        let jsonData = try JSONEncoder().encode(config)
        guard let jsonString = String(data: jsonData, encoding: .utf8) else {
            throw CausalityError.encoding("Failed to encode config")
        }
        print("[Causality:Bridge] Init JSON: \(jsonString)")
        let result = CAUMobileInit(jsonString)
        print("[Causality:Bridge] Init result: '\(result)'")
        if !result.isEmpty {
            throw CausalityError.initialization(result)
        }
    }

    static func track(event: Event) throws {
        let jsonData = try JSONEncoder().encode(event)
        guard let jsonString = String(data: jsonData, encoding: .utf8) else {
            throw CausalityError.encoding("Failed to encode event")
        }
        print("[Causality:Bridge] Track JSON: \(jsonString)")
        let result = CAUMobileTrack(jsonString)
        print("[Causality:Bridge] Track result: '\(result)'")
        if !result.isEmpty {
            throw CausalityError.tracking(result)
        }
    }

    static func trackTyped<E: CausalityEvent>(event: E) throws {
        let propsData = try JSONEncoder().encode(event)
        guard let propsJSON = String(data: propsData, encoding: .utf8) else {
            throw CausalityError.encoding("Failed to encode event properties")
        }
        let jsonString = "{\"type\":\"\(E.eventType)\",\"properties\":\(propsJSON)}"
        print("[Causality:Bridge] Track JSON: \(jsonString)")
        let result = CAUMobileTrack(jsonString)
        print("[Causality:Bridge] Track result: '\(result)'")
        if !result.isEmpty {
            throw CausalityError.tracking(result)
        }
    }

    static func setUser(userId: String, traits: [String: AnyCodable]?, aliases: [String]?) throws {
        struct UserPayload: Codable {
            let userId: String
            let traits: [String: AnyCodable]?
            let aliases: [String]?

            enum CodingKeys: String, CodingKey {
                case userId = "user_id"
                case traits
                case aliases
            }
        }

        let payload = UserPayload(userId: userId, traits: traits, aliases: aliases)
        let jsonData = try JSONEncoder().encode(payload)
        guard let jsonString = String(data: jsonData, encoding: .utf8) else {
            throw CausalityError.encoding("Failed to encode user")
        }
        print("[Causality:Bridge] SetUser JSON: \(jsonString)")
        let result = CAUMobileSetUser(jsonString)
        print("[Causality:Bridge] SetUser result: '\(result)'")
        if !result.isEmpty {
            throw CausalityError.identification(result)
        }
    }

    static func reset() throws {
        print("[Causality:Bridge] Reset called")
        let result = CAUMobileReset()
        print("[Causality:Bridge] Reset result: '\(result)'")
        if !result.isEmpty {
            throw CausalityError.reset(result)
        }
    }

    static func resetAll() throws {
        print("[Causality:Bridge] ResetAll called")
        let result = CAUMobileResetAll()
        print("[Causality:Bridge] ResetAll result: '\(result)'")
        if !result.isEmpty {
            throw CausalityError.reset(result)
        }
    }

    static func flush() throws {
        print("[Causality:Bridge] Flush called")
        let result = CAUMobileFlush()
        print("[Causality:Bridge] Flush result: '\(result)'")
        if !result.isEmpty {
            throw CausalityError.flush(result)
        }
    }

    static func getDeviceId() -> String {
        let result = CAUMobileGetDeviceId()
        print("[Causality:Bridge] GetDeviceId: '\(result)'")
        return result
    }

    static func isInitialized() -> Bool {
        let result = CAUMobileIsInitialized()
        print("[Causality:Bridge] IsInitialized: \(result)")
        return result
    }

    static func appDidEnterBackground() {
        print("[Causality:Bridge] AppDidEnterBackground called")
        _ = CAUMobileAppDidEnterBackground()
    }

    static func appWillEnterForeground() {
        print("[Causality:Bridge] AppWillEnterForeground called")
        _ = CAUMobileAppWillEnterForeground()
    }
}

File: sdk/android/causality/src/main/kotlin/io/causality/internal/Bridge.kt
package io.causality.internal

import io.causality.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import mobile.Mobile

internal object Bridge {
    private val json = Json {
        encodeDefaults = false
        ignoreUnknownKeys = true
    }

    fun initSDK(config: Config) {
        val configJson = json.encodeToString(config)
        val result = Mobile.init(configJson)
        if (result.isNotEmpty()) {
            throw CausalityException.Initialization(result)
        }
    }

    fun track(event: Event) {
        val eventJson = json.encodeToString(event)
        val result = Mobile.track(eventJson)
        if (result.isNotEmpty()) {
            throw CausalityException.Tracking(result)
        }
    }

    @Serializable
    private data class UserPayload(
        @kotlinx.serialization.SerialName("user_id") val userId: String,
        val traits: Map<String, String>? = null,
        val aliases: List<String>? = null
    )

    fun setUser(userId: String, traits: Map<String, String>?, aliases: List<String>?) {
        val payload = UserPayload(userId, traits, aliases)
        val payloadJson = json.encodeToString(payload)
        val result = Mobile.setUser(payloadJson)
        if (result.isNotEmpty()) {
            throw CausalityException.Identification(result)
        }
    }

    fun reset() {
        val result = Mobile.reset()
        if (result.isNotEmpty()) {
            throw CausalityException.Reset(result)
        }
    }

    fun resetAll() {
        val result = Mobile.resetAll()
        if (result.isNotEmpty()) {
            throw CausalityException.Reset(result)
        }
    }

    fun flush() {
        val result = Mobile.flush()
        if (result.isNotEmpty()) {
            throw CausalityException.Flush(result)
        }
    }

    fun getDeviceId(): String = Mobile.getDeviceId()

    fun isInitialized(): Boolean = Mobile.isInitialized()

    fun appDidEnterBackground() {
        Mobile.appDidEnterBackground()
    }

    fun appWillEnterForeground() {
        Mobile.appWillEnterForeground()
    }
}
